optimum/neuron/distributed/decoder_models.py

```diff
@@ -130,6 +132,150 @@ def _parallelize(
         return model
 
 
+class GPTNeoXParallelEmbedding(ParallelEmbedding):
+    EMBEDDING_NAME = "embed_in"
+    LM_HEAD_NAME = "embed_out"
+
+
+class GPTNeoXParallelSelfAttention(ParallelSelfAttentionWithFusedQKV):
+    QUERY_KEY_VALUE_NAME = "query_key_value"
+    OUTPUT_PROJECTION_NAME = "dense"
+    NUM_ATTENTION_HEADS_NAME = "num_attention_heads"
+    ALL_HEAD_SIZE_NAME = "hidden_size"
+
+
+class GPTNeoXParallelMLP(ParallelMLP):
+    FIRST_LINEAR_NAME = "dense_h_to_4h"
+    SECOND_LINEAR_NAME = "dense_4h_to_h"
+
+
+class GPTNeoXParallelCrossEntropy(ParallelCrossEntropy):
+    LAST_LINEAR_PROJECTION_NAME = "embed_out"
+
+
+class GPTNeoXParallelizer(Parallelizer):
+    SEQUENCE_PARALLEL_LAYERNORM_PATTERNS = [
+        "gpt_neox.layers.[0-9]+.input_layernorm",
+        "gpt_neox.layers.[0-9]+.post_attention_layernorm",
+        "gpt_neox.final_layer_norm",
+    ]
+    SEQUENCE_COLLECTIVE_OPS_INFOS = [
+        SequenceCollectiveOpInfo("scatter", torch.nn.Embedding, "output", "first"),
+        SequenceCollectiveOpInfo("gather", torch.nn.LayerNorm, "output", "last"),
+    ]
+
+    @classmethod
+    def patch_for_sequence_paralelism(cls, model: "PreTrainedModel", sequence_parallel_enabled: bool):
+        if not sequence_parallel_enabled:
+            return
+
+        def sequence_parallel_forward(
+            self,
+            hidden_states: torch.FloatTensor,
+            attention_mask: torch.FloatTensor,
+            position_ids: torch.LongTensor,
+            head_mask: Optional[torch.FloatTensor] = None,
+            layer_past: Optional[Tuple[torch.Tensor]] = None,
+            use_cache: Optional[bool] = False,
+            output_attentions: Optional[bool] = False,
+        ):
+            has_layer_past = layer_past is not None
+
+            # Compute QKV
+            # Attention heads [batch, seq_len, hidden_size]
```

Comment:
If sequence_parallel is enabled, it is actually `[seq_len, batch, ... ] `right ?

Reply:
Yes, same I copy pasted. Will change this.